[20.01] Check dataset is accessible and use payload in job error API

nsoranzo opened the pull request:
Without the `payload` parameter, it cannot be correctly used by BioBlend.

Also:
- `self.decode_id()` already raise a `MalformedId` exception
- Use `RequestParameterMissingException` instead of `ObjectAttributeMissingException`  when no object is to be created
- Prevent a `KeyError` exception if `dataset_id` is missing from the payload
- Check that the job is in error and the dataset is an output of the job

nsoranzo committed:
Check dataset is accessible and use payload in job error API

Without the `payload` parameter, it cannot be correctly used by BioBlend.

Also:
- `self.decode_id()` already raise a `MalformedId` exception
- Use `RequestParameterMissingException` instead of
  `ObjectAttributeMissingException`  when no object is to be created
- Prevent a `KeyError` exception if `dataset_id` is missing from the
  payload
- Check that the dataset is an output of the job

nsoranzo commented on lib/galaxy_test/api/test_jobs.py:
```diff
@@ -243,9 +244,9 @@ def test_report_error_anon(self):
         job_id = run_response['jobs'][0]["id"]
```
Since I've introduced a check in the `JobController.error()` API to verify that the job is in error, I need to add `self.dataset_populator.wait_for_job(job_id)` here, but that fails with:
```
urllib3.connectionpool DEBUG 2020-05-12 12:26:41,891 Starting new HTTP connection (1): 127.0.0.1:8506
galaxy.web.framework.decorators ERROR 2020-05-12 12:26:41,925 Uncaught exception in exposed API method:
Traceback (most recent call last):
  File "/usr/users/ga002/soranzon/software/nsoranzo_galaxy/lib/galaxy/web/framework/decorators.py", line 282, in decorator
    rval = func(self, trans, *args, **kwargs)
  File "/usr/users/ga002/soranzon/software/nsoranzo_galaxy/lib/galaxy/webapps/galaxy/api/jobs.py", line 144, in show
    job = self.__get_job(trans, id)
  File "/usr/users/ga002/soranzon/software/nsoranzo_galaxy/lib/galaxy/webapps/galaxy/api/jobs.py", line 366, in __get_job
    return self.job_manager.get_accessible_job(trans, decoded_job_id)
  File "/usr/users/ga002/soranzon/software/nsoranzo_galaxy/lib/galaxy/managers/jobs.py", line 58, in get_accessible_job
    belongs_to_user = (job.user == trans.user) if job.user else (job.session_id == trans.get_galaxy_session().id)
AttributeError: 'NoneType' object has no attribute 'id'
urllib3.connectionpool DEBUG 2020-05-12 12:26:41,926 http://127.0.0.1:8506 "GET /api/jobs/0beccb6ec965c2fd?full=False&key=46459b20d08abf80d32d647860e67069 HTTP/1.1" 500 None
```
because it uses the test user API key not the anonymous user cookie.

Should I:
- drop this API test? (we don't have many other API tests for anonymous users any way)
- or drop the check that the job is in error?

Extending the test populators to support anonymous session seems to be a much larger project. Suggestions?

jmchilton replied:
I'd comment out the check that the job is in error until we can provide some support in populators for doing more with anonymous testing. We don't have much anonymous user testing but that is sort of a reason to not drop the tests we have in my mind.

If instead of you decide to just drop the test though I won't be -1 or anythinig... just in my mind the test is more valuable than the job in error check.

mvdbeek replied:
We shouldn't remove this test, I added it because anonymous error reporting was broken. I think we may want to allow submitting reports for all job states (say someone wants to report an empty dataset, or a job that hangs in the queue or ...). Not that this is exposed in the UI ... but I don't think the state check is necessary.

nsoranzo replied:
Thanks both for the feedback, I've removed the check that the job is in error, but kept the `self.dataset_populator.wait_for_job(job_id)` call in `test_report_error()` above.

mvdbeek commented on lib/galaxy_test/api/test_jobs.py:
```diff
@@ -243,9 +244,9 @@ def test_report_error_anon(self):
+                                 data={'email': 'someone@domain.com', 'dataset_id': dataset_id},
                                  cookies=cookies)
-        assert response.status_code == 200
+        assert response.status_code == 200, response.json()
```
```suggestion
        assert response.status_code == 200, response.text
```
It's very likely that non-200 status codes won't be valid JSON.

nsoranzo replied:
This worked fine in my testing:
```
self = <galaxy_test.api.test_jobs.JobsApiTestCase testMethod=test_report_error_anon>

    @skip_without_tool('detect_errors_aggressive')
    def test_report_error_anon(self):
        # Need to get a cookie and use that for anonymous tool runs
        cookies = requests.get(self.url).cookies
        payload = json.dumps({"tool_id": "detect_errors_aggressive",
                              "inputs": {"error_bool": "true"}})
        run_response = requests.post("%s/tools" % self.galaxy_interactor.api_url, data=payload, cookies=cookies).json()
        job_id = run_response['jobs'][0]["id"]
        dataset_id = run_response['outputs'][0]['id']
        response = requests.post('%s/jobs/%s/error' % (self.galaxy_interactor.api_url, job_id),
                                 data={'email': 'someone@domain.com', 'dataset_id': dataset_id},
                                 cookies=cookies)
>       assert response.status_code == 200, response.text
E       AssertionError: {"err_msg": "Job is not in error", "err_code": 400008}
E       assert 400 == 200
E         -400
E         +200

lib/galaxy_test/api/test_jobs.py:248: AssertionError
```

and is used already in `lib/galaxy_test/api/test_search.py` and `lib/galaxy_test/api/test_tools.py`

mvdbeek replied:
Make it a 500 and it will probably not work. I don't think it's a good idea to assume any response will be in JSON if you don't see the expect status code.

mvdbeek replied:
And we should change those other instances too. It's a minor thing, but it isn't quite right.

nsoranzo committed:
Use `response.text` instead of `response.json()` when request failed

We cannot be sure that the response is valid JSON if there was an error.

Also do not use mutable data structures for argument defaults.

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>
